subsys/bluetooth/services/hrs_client.c

```diff
@@ -0,0 +1,414 @@
+
+	if (err) {
+		LOG_ERR("Read value error: %d", err);
+		err_cb = err;
+	} else if (length != sizeof(uint8_t)) {
+		err_cb = -EMSGSIZE;
+	} else {
+		location = *((uint8_t *)data);
+	}
+
+	read_cb = hrs_c->sensor_location_char.read_cb;
+	atomic_clear_bit(&hrs_c->state, HRS_SENSOR_LOCATION_READ_IN_PROGRES);
+
+	if (read_cb) {
+		read_cb(hrs_c, (enum bt_hrs_client_sensor_location)location, err_cb);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static uint8_t on_hrs_measurement_notify(struct bt_conn *conn,
+					 struct bt_gatt_subscribe_params *params,
+					 const void *data, uint16_t length)
+{
+	int err;
+	struct bt_hrs_client *hrs_c;
+	struct bt_hrs_client_measurement hr_measurement;
+
+	hrs_c = CONTAINER_OF(params, struct bt_hrs_client, measurement_char.notify_params);
+
+	if (!data) {
+		atomic_clear_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED);
+		LOG_DBG("[UNSUBSCRIBE] from Heart Rate Measurement characterictic");
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	LOG_HEXDUMP_DBG(data, length, "[NOTIFICATION] Heart Rate Measurement: ");
+
+	err = hrs_measurement_data_parse(&hr_measurement, data, length);
+
+	if (hrs_c->measurement_char.notify_cb) {
+		hrs_c->measurement_char.notify_cb(hrs_c, &hr_measurement, err);
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void on_cp_write(struct bt_conn *conn, uint8_t err,
+					struct bt_gatt_write_params *params)
+{
+	struct bt_hrs_client *hrs_c;
+	bt_hrs_client_write_cb write_cb;
+
+	hrs_c = CONTAINER_OF(params, struct bt_hrs_client, cp_char.write_params);
+
+	write_cb = hrs_c->cp_char.write_cb;
+	atomic_clear_bit(&hrs_c->state, HRS_CONTROL_POINT_WRITE_PENDING);
+
+	if (write_cb) {
+		write_cb(hrs_c, err);
+	}
+}
+
+int bt_hrs_client_measurement_subscribe(struct bt_hrs_client *hrs_c,
+					bt_hrs_client_notify_cb notify_cb)
+{
+	int err;
+	struct bt_gatt_subscribe_params *params = &hrs_c->measurement_char.notify_params;
+
+	if (!hrs_c || !notify_cb) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_and_set_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED)) {
+		LOG_ERR("Heart Rate Measurement characterisic notifications already enabled.");
+		return -EALREADY;
+	}
+
+	hrs_c->measurement_char.notify_cb = notify_cb;
+
+	params->ccc_handle = hrs_c->measurement_char.ccc_handle;
+	params->value_handle = hrs_c->measurement_char.handle;
+	params->value = BT_GATT_CCC_NOTIFY;
+	params->notify = on_hrs_measurement_notify;
+
+	atomic_set_bit(params->flags, BT_GATT_SUBSCRIBE_FLAG_VOLATILE);
+
+	err = bt_gatt_subscribe(hrs_c->conn, params);
+	if (err) {
+		atomic_clear_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED);
+		LOG_ERR("Subscribe to Heart Rate Measurement characteristic failed");
+	} else {
+		LOG_DBG("Subscribed to Heart Rate Measurement characteristic");
+	}
+
+	return err;
+}
+
+int bt_hrs_client_measurement_unsubscribe(struct bt_hrs_client *hrs_c)
+{
+	int err;
+
+	if (!hrs_c) {
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(&hrs_c->state, HRS_MEASUREMENT_NOTIFY_ENABLED)) {
+		return -EFAULT;
```

Comment:
Shouldn't the error code be symmetric here to the one used in the bt_hrs_client_measurement_subscribe function?

-EALREADY

Reply:
This error code is returned by our other service clients do to be consistent I didn't change it